packages/pymedphys_pinnacle/tests/test_pinnacle.py

```diff
@@ -0,0 +1,92 @@
+import pytest
+import pydicom
+import numpy
+import os
+import shutil
+import tempfile
+
+DATA_DIRECTORY = os.path.join(
+    os.path.dirname(__file__), "data")
+
+pinn_archive = os.path.join(DATA_DIRECTORY, 'pinn/test_pinnacle_16.0.tar.gz')
+pinn_path = 'Institution_2812/Mount_0/Patient_16218'
+gt_dose_path = os.path.join(
+    DATA_DIRECTORY, 'dcm/1.3.46.670589.13.997910418.20181018190715.183309.dcm')
+working_path = tempfile.mkdtemp()
+archive_path = os.path.join(working_path, 'pinn')
+export_path = os.path.join(working_path, 'output')
+
+
+@pytest.fixture(scope='session')
+def archive():
+    import tarfile
+
+    shutil.rmtree(working_path, ignore_errors=True)
+    os.makedirs(working_path)
+
+    t = tarfile.open(pinn_archive)
+
+    for m in t.getmembers():
+        if not ':' in m.name:
+            t.extract(m, path=archive_path)
+    return os.path.join(archive_path, pinn_path)
+
+
+@pytest.fixture
+def pinn(archive):
+    from pymedphys_pinnacle import Pinnacle
+    return Pinnacle(archive, None)
+
+
+def test_pinnacle(pinn):
+    plans = pinn.get_plans()
+    assert len(plans) == 3
+
+
+def test_dose(pinn):
```

Comment:
During development, I use a script which simply outputs all of the differences between the two DICOM files. However I don't think this is useful for tests at this stage, since there are going to be a lot of differences. For now we just want to make sure the stuff we're interested in is matching I think. Also for the contour data, the sequence of points might be different, but that doesn't mean its necessarily wrong.

Reply:
A straight equality check will be quite difficult.

You've probably already thought of something like this, but might there be value in creating a mask based off of the contours and then verifying that the masks agree for each named contour?

Reply:
Yep, that's exactly the approach I was planning to take... So... next question, is there a function in pymedphys which will generate a mask from an RTStruct for me? :-P

Reply:
Something like this :)

https://github.com/pymedphys/pymedphys/blob/master/packages/pymedphys_dicom/src/pymedphys_dicom/dicom/dose.py#L187-L219

Reply:
It's not exactly right...

Reply:
That masks a dose grid. In reality the "masking" feature should be refactored into it's own function.

Reply:
Awesome, yeah that's pretty close to what I'm after! :)

Reply:
Why is a straight equality check difficult? It's already implemented in `pydicom`.

Just to be clear, I'm suggesting setting a well-established baseline and running simple equality test for constancy.

Reply:
In this case I don't think a baseline will be anywhere near as helpful as comparing to an available ground truth. Take a true DICOM export from Pinnacle and using that as the ground truth compare the created DICOMs to it. That will do the job of a baseline while also actually testing it works (not just testing that it is the same as it was before).

Reply:
Agree that comparison with ground truth is most important.

But if for some reason full equality cannot or will not be achieved (e.g. missing info in the .tar file or even developer discretion), a baseline check is still valuable to protect against the effect of future code changes that could have adverse consequences outside of the ground truth test scope.

If full equality CAN be achieved against ground truth, then it is implicitly a full baseline check :)

My thoughts were: if @pchlap is happy with the performance of the converter from his own experience and manual testing, but doesn't have the auto tests written, perhaps a baseline test is a satisfactory minimum to include prior to merge?

But if you prefer the stronger constraint of automated tests verifying against ground truth with some minimum level of comprehensiveness, I certainly won't object :)

Reply:
In this case the ground truth is readily accessible. And I would argue a complete equality with ground truth isn't required. For example certain headers can be left out. Order of things can be different as long as meaning is the same.

A forced complete baseline has the issue of forcing some things to stay the same which shouldn't need to stay the same. Which can be problematic. I think @pchlap approach here is ideal for this case

Reply:
I disagree.

A completely test enforces nothing. It simply flags (possibly unintended) changes. A developer can simply alter the test to accommodate. Thus is best practice. A catch-all is prudent.

Reply:
Okay... But should it catch that "GE Medical Systems" is hard coded into the header?

Reply:
And should it catch that contour A is saved in the file before contour B?

Reply:
Test specificity is important too. If tests often fail when no code logic is broken, then we will get "test failing fatigue". It's okay for tests to throw false negatives, but if from the outset we can catch just as many true positives with less false negatives I'm keen for that approach.

If a test is brittle, and fails because the order of something changed, and that happens on multiple tests throughout the code base, that hampers the efficiency of important refactoring work. Tests are supposed to make refactoring safer and more efficient. If the design of tests make refactoring painful, because say, "we don't want to run this for loop in a different order, as that will break 10 tests", we just leave the for loop as is.

Baselines are certainly important. I do agree. But it's baselines on the logic and meaningful outcomes. Things such as order, noise, non-determinism, non-logically relevant equalities should be kept out of the baselines as much as is plausible.

An example, if a swathe of tests relied on something such as dictionary key ordering, then when upgrading from Python 3.5 --> 3.6 all those tests would have then failed... (https://stackoverflow.com/a/39980744/3912576) ... which would have been very painful.

Reply:
> Okay... But should it catch that "GE Medical Systems" is hard coded into the header?

Actually, yes, it absolutely should! If this has changed, we should be aware and adjust the test accordingly. I think the problem you envisage here is not solved by omitting it from testing, but by not hard coding it in the first place...

> And should it catch that contour A is saved in the file before contour B?

No, but that shouldn't be caught by the equality operator anyway.

I think we roughly agree on the endpoint, but differ on the means to get there. I agree that you shouldn't write tests to be too strict. You shouldnt require a certain order when order is known to be unimportant. My point is that you ONLY want to omit *known* false positives from your testing. I'm saying that one ought to test with maximum coverage and then narrow down from there. E.g. if your initial, very strictly written test fails for something like dict key order (it shouldnt anyway, but we'll use it as an example), then adjust your test to accommodate different key orders and iterate from there. That way, you are only ever accommodating known things, you are far less likely to miss unintended adverse changes as compared to the bottom-up, "only write unit tests for things you can think of" approach.

Reply:
> That way, you are only ever accommodating known things, you are far less likely to miss unintended adverse changes as compared to the bottom-up, "only write unit tests for things you can think of" approach.

Fair point. I agree.

> I think the problem you envisage here is not solved by omitting it from testing, but by not hard coding it in the first place...

Also fair point.

Yup.

Reply:
> No real objection to this approach; plenty of good sanity checks here! Though would a simple test of equality for your extracted DICOMs vs expected DICOMs be sufficient? Indeed perhaps superior, since all tags are checked?

@Matthew-Jennings for this request to use a straight equality to be useful to @pchlap I think you might have to provide a DICOM structure equality test example that still evaluates to `True` even if the structure slices are out of order, or if the structure points are defined anti-clockwise as opposed to clockwise for example. Essentially, if the structure definitions define the same volume this said equality check would need to evaluate to `True`.

I'm asking this... Because I think the request is non-trivial. And I believe that @pchlap current trajectory approach is sufficient.

Reply:
I still disagree, pending evidence that such an implementation *is* required for this request. I.e. writing my suggested simple test and demonstrating that it would actually result in an unwieldy number of false positive failures. If written intelligently (yet still trivial), I’m not so sure it would... :)

Unfortunately I won’t be in a position to do this myself for at least another month!

That said, I don’t mean to suggest that I would wish to block this merge for this request (note my original comment)! Most of this is academic. Though I would prefer if my suggestion were at least *tried*, please feel free to proceed. :)

Reply:
Remember I’m suggesting relative to baseline, not ground truth (so in *addition* to the existing tests)